### DISCUSSION TITLE
Installing npm packages using bun

### DISCUSSION THREAD

**Discussion by jay-babu** [mason-org/mason.nvim#1031] [General]
Hello,

https://github.com/williamboman/mason.nvim/blob/main/lua/mason-registry/index/typescript-language-server/init.lua. tsserver is installed using npm, which is probably preferred as it is stable. However there have been reports of big performance gains when using tsserver installed via bun, [see reddit post](https://www.reddit.com/r/neovim/comments/11bpzmk/lsp_typescript_and_tailwind_with_bun_major/?utm_source=share&utm_medium=ios_app&utm_name=iossmf). Is there a way to get npm packages installed via bun instead while still using mason?

**Comment by williamboman** [comment]
Hey, interesting! So installing "npm" packages via `bun` will not be considered with the rationale that installation of packages are normally done pretty infrequently via Mason. Furthermore all installations of npm packages are very simple (1 or 2 packages at most). As such, the room for meaningful perf gains from using Bun for package installation is relatively small.

As for running installed packages via `bun` instead of `node`, I think you should be able to do something like

```lua
lspconfig.tsserver.setup {
  cmd = { "bun", "run", "--bun", "typescript-language-server", "--stdio" }
}
```

This should execute the `typescript-language-server` executable provided by Mason in the Bun runtime.

I don't believe Bun has full parity with Node.js yet, so use at own risk!

**Reply by williamboman** [reply]
I _think_ I'd be open for a PR to `mason-lspconfig.nvim` that adds this a setting and modifies CMDs accordingly (my only qualm is that I think such a feature would more appropriately belong in nvim-lspconfig, but I doubt they'd want to include something like that).

**Reply by williamboman** [reply]
Actually scratch that, adding it to mason-lspconfig.nvim would be too much of a scope creep. This would be a pretty easy plugin <sup>(*cough* `bunify.nvim`)</sup> to write though:

```lua
local util = require "lspconfig.util"

local function prefix_bun(cmd)
    return vim.list_extend({
        "bun",
        "run",
        "--bun",
    }, cmd)
end

util.on_setup = util.add_hook_before(util.on_setup, function(config, user_config)
    if config.cmd and is_node_server(config.name) then
        config.cmd = prefix_bun(config.cmd)
    end
end)
```

Note that the above probably will be problematic on Windows because how lspconfig [prefixes](https://github.com/williamboman/nvim-lspconfig/blob/cb64ead000d70716cd779ef0409ebc2ce4a464fd/lua/lspconfig/server_configurations/jsonls.lua#L6-L8) these cmd's with `cmd.exe /C`.

**Comment by sljeff** [comment]
I tried to modify the lsp cmd, but the compatibility of bun is so poor that typescript-language-server cannot run successfully.

**Comment by Senal-D-A-Gunaratna** [comment]
Fine, I'll do it myself
@jay-babu  asked upstream for this back in 2023 — mason maintainers passed, reasonably, given bun's Node compat wasn't there yet and npm installs are infrequent anyway. Two years and a much better bun later, I built [swapson.nvim](https://github.com/Senal-D-A-Gunaratna/swapson.nvim) — it goes further than what was originally asked, fully replacing mason's npm installer with bun, not just the runtime.
Works on Linux (tested), should work on macOS.
Windows? Absolutely not.

**Reply by danila-schelkov** [reply]
Thanks!
